<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Paulo Czarnewski" />

<meta name="date" content="2019-09-10" />

<title>PBMC Tutorial</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4,h5",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Sauron</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="pbmc_tutorial.html">Tutorial</a>
</li>
<li>
  <a href="functions.html">Functions</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">PBMC Tutorial</h1>
<h4 class="author">Paulo Czarnewski</h4>
<h4 class="date">2019-09-10</h4>

</div>


<style type="text/css">

body{ /* Normal  */
      font-size: 14px;
  }
td {  /* Table  */
  font-size: 14px;
}
h1.title {
  font-size: 24px;
  color: DarkBlue;
}
h1 { /* Header 1 */
  font-size: 20px;
  color: DarkRed;
}
h2 { /* Header 2 */
    font-size: 16px;
}
h3 { /* Header 3 */
  font-size: 18px;
  font-family: "Times New Roman", Times, serif;
}
code.r{ /* Code block */
    font-size: 14px;
}
pre { /* Code block - determines code spacing between lines */
    font-size: 14px;
}
</style>
<p><br/></p>
<p>This is a basic guided analysis for using 3 PBMC datasets</p>
<p><br/></p>
<div id="install-software-dependecies" class="section level1">
<h1>Install software dependecies</h1>
<p>The only two softwares needed are <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/"><strong>Conda</strong></a> and <a href="https://www.atlassian.com/git/tutorials/install-git"><strong>git</strong></a>. They should be installed using their instructions.</p>
<p>All other software used in Sauron is managed by Conda environment and will be installed automatically. These includes <strong>R</strong> , <strong>Rstudio</strong> , <strong>Python</strong> and all necessary packages / libraries to run the workflow. The complete list with software and their versions can be found in the <code>environment.yml</code> file.</p>
<p><br/></p>
</div>
<div id="clone-this-repository" class="section level1">
<h1>Clone this repository</h1>
<p>First, you will need to clone this repo into your project folder. For this tutorial we will create a folder <code>sauron_tutorial_PBMC</code> inside our <code>Downloads</code> folder:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="fu">mkdir</span> ~/Downloads/sauron_tutorial_PBMC</span>
<span id="cb1-2"><a href="#cb1-2"></a></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="bu">cd</span> ~/Downloads/sauron_tutorial_PBMC</span>
<span id="cb1-4"><a href="#cb1-4"></a></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">git</span> clone https://czarnewski@bitbucket.org/scilifelab-lts/sauron.git</span></code></pre></div>
<p>This will create a folder named “sauron” in your project folder, contating all the files required for the analysis.</p>
<p>Alternativelly, you can also simply create these folders and download the repository manually. Your folder structure should look like this:</p>
<p><img src="figures/sauron_folder.png" style="width:40.0%" /></p>
<p><br/></p>
</div>
<div id="download-pbmc-dataset" class="section level1">
<h1>Download PBMC dataset</h1>
<p>Here, we will download 3 PBMC datasets to work with. All datasets will be put into the <code>data</code> folder. Each dataset should be placed individually within its own folder. This enforces the user to have well confided raw data condicent with good data management practices. This further allows designing ver complex designs using the metadata table.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="bu">cd</span> ~/Downloads/sauron_tutorial_PBMC/sauron/data</span>
<span id="cb2-2"><a href="#cb2-2"></a></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="fu">mkdir</span> pbmc_1k_v2</span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="ex">curl</span> -o pbmc_1k_v2/pbmc_1k_v2_filtered_feature_bc_matrix.h5 -O \</span>
<span id="cb2-5"><a href="#cb2-5"></a>http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v2/pbmc_1k_v2_filtered_feature_bc_matrix.h5</span>
<span id="cb2-6"><a href="#cb2-6"></a></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="fu">mkdir</span> pbmc_1k_v3</span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="ex">curl</span> -o pbmc_1k_v3/pbmc_1k_v3_filtered_feature_bc_matrix.h5 -O \</span>
<span id="cb2-9"><a href="#cb2-9"></a>http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_filtered_feature_bc_matrix.h5</span>
<span id="cb2-10"><a href="#cb2-10"></a></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="fu">mkdir</span> pbmc_1k_p3</span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="ex">curl</span> -o pbmc_1k_p3/pbmc_1k_protein_v3_filtered_feature_bc_matrix.h5 -O \</span>
<span id="cb2-13"><a href="#cb2-13"></a>http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_protein_v3/pbmc_1k_protein_v3_filtered_feature_bc_matrix.h5</span></code></pre></div>
<p><br/></p>
<hr />
</div>
<div id="adding-metadata-to-your-folder" class="section level1">
<h1>Adding metadata to your folder</h1>
<p>Add your data and metadata into the <code>data</code> directory. One dataset matrix per folder (i.e. one plate per folder or one 10X lane per folder). Name each folder as the desired sample names. The sample names should match the names in the 1st column of you metadata csv file.</p>
<p>We will manually add information into the metadata file (which can be created with any spread sheet editor) and saved as <code>.csv</code>.</p>
<p><img src="figures/metadata_file.png" style="width:40.0%" /></p>
<p>It is important to notice that each line corresponds to a dataset and the <strong>first column</strong> is exactly the name of the dataset folders. Only names found in both the metadata and in the data folder will be used.</p>
<p><br/></p>
<p>Your final folder should look like this:</p>
<p><img src="figures/data_folder.png" style="width:50.0%" /></p>
<p><br/></p>
<hr />
</div>
<div id="loading-the-data" class="section level1">
<h1>Loading the data</h1>
<p>With the data in place, we can now define some anlaysis parameters and then load it to Sauron. Here we will define some metadata parameters we would like to use for plotting later on. Those are defined based on the column names of the metadata file above.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="va">var_to_plot=</span><span class="st">&#39;assay,chemistry&#39;</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="va">var_to_regress=</span><span class="st">&#39;nFeature_RNA,percent_mito,S.Score,G2M.Score&#39;</span></span></code></pre></div>
<p>We can now load them using the <code>00_load_data.R</code> function.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="ex">Rscript</span> <span class="va">$script_path</span>/00_load_data.R \</span>
<span id="cb4-2"><a href="#cb4-2"></a>--input_path <span class="va">$main</span>/<span class="st">&#39;data&#39;</span> \</span>
<span id="cb4-3"><a href="#cb4-3"></a>--dataset_metadata_path <span class="va">$main</span>/<span class="st">&#39;data/metadata.csv&#39;</span> \</span>
<span id="cb4-4"><a href="#cb4-4"></a>--assay <span class="st">&#39;rna&#39;</span> \</span>
<span id="cb4-5"><a href="#cb4-5"></a>--output_path <span class="va">$main</span>/<span class="st">&#39;analysis/1_qc&#39;</span> \</span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="op">2&gt;&amp;1</span> <span class="kw">|</span> <span class="fu">tee</span> <span class="va">$main</span>/log/<span class="st">&#39;00_load_data_log.txt&#39;</span></span></code></pre></div>
<p>This will create a seurat object with a slot named <code>rna</code> containing all the counts. The results will be outputed in the folder <code>analysis/1_qc</code>. The log file after running this function can be seen in the log folder <code>log/00_load_data_log.txt</code>. There should be now a <code>raw_seurat_object.rds</code> in your qc folder..</p>
<p><img src="figures/loaded_object.png" style="width:45.0%" /></p>
<p><br/></p>
<hr />
</div>
<div id="quality-control" class="section level1">
<h1>Quality control</h1>
<p>Once the data is loaded into the Seurat object, it becomes easy to work with it and perform several quality control (QC) metris using the function <code>01_qc_filter.R</code>. By default, this function will compute cell cycle scoring, remove non-coding genes (including pseudogenes), calculate percentage of several gene families.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="ex">Rscript</span> <span class="va">$script_path</span>/01_qc_filter.R \</span>
<span id="cb5-2"><a href="#cb5-2"></a>--Seurat_object_path <span class="va">$main</span>/<span class="st">&#39;analysis/1_qc/raw_seurat_object.rds&#39;</span> \</span>
<span id="cb5-3"><a href="#cb5-3"></a>--columns_metadata <span class="va">$var_to_plot</span> \</span>
<span id="cb5-4"><a href="#cb5-4"></a>--species_use <span class="st">&#39;hsapiens&#39;</span> \</span>
<span id="cb5-5"><a href="#cb5-5"></a>--remove_non_coding <span class="st">&#39;True&#39;</span> \</span>
<span id="cb5-6"><a href="#cb5-6"></a>--plot_gene_family <span class="st">&#39;RPS,RPL,mito,HB&#39;</span> \</span>
<span id="cb5-7"><a href="#cb5-7"></a>--remove_gene_family <span class="st">&#39;mito&#39;</span> \</span>
<span id="cb5-8"><a href="#cb5-8"></a>--min_gene_count <span class="st">&#39;5&#39;</span> \</span>
<span id="cb5-9"><a href="#cb5-9"></a>--min_gene_per_cell <span class="st">&#39;200&#39;</span> \</span>
<span id="cb5-10"><a href="#cb5-10"></a>--assay <span class="st">&#39;rna&#39;</span> \</span>
<span id="cb5-11"><a href="#cb5-11"></a>--output_path <span class="va">$main</span>/analysis/1_qc \</span>
<span id="cb5-12"><a href="#cb5-12"></a><span class="op">2&gt;&amp;1</span> <span class="kw">|</span> <span class="fu">tee</span> <span class="va">$main</span>/log/<span class="st">&#39;01_QC_log.txt&#39;</span></span></code></pre></div>
<p>By default, this function will calculate several percentage estimates and output two kind of files, either contating ALL/RAW cells or FILTERED cells only. Thus, you can later choose which to proceed if you dont want to exclude any cell at first. Your folder should look like this:</p>
<p><img src="figures/QC_folder.png" style="width:40.0%" /></p>
<p>By default, cells need to be within:</p>
<ul>
<li>0 - 25 % mitocondrial genes</li>
<li>0 - 50 % RPS genes</li>
<li>0 - 50 % RPL genes</li>
<li>0.5 - 99.5 quantiles of number UMIs (removing extreme ouliers)</li>
<li>0.5 - 99.5 quantiles of number of counts (removing extreme ouliers)</li>
<li>90 - 100 % protein coding genes</li>
<li>.9 - 1 Gini index</li>
<li>.95 - 1 Simpson index</li>
</ul>
<p>Some of the outputed plots can be found below, before and after filtering:</p>
<details>
<p><strong>ALL cells</strong></p>
<p><img src="figures/QC_dataset_ALL.png" /></p>
<p><br/></p>
<p><strong>FILTERED cells only</strong></p>
<p><img src="figures/QC_dataset_FILTERED.png" /></p>
</details>
<p><br/></p>
<hr />
</div>
<div id="dataset-integration" class="section level1">
<h1>Dataset integration</h1>
<p>Once low quality cells were filtered out we can proceed to integrate the datasets with the function <code>02_integrate.R</code>. It will adjust for batch effects if present and output into a integrated space in the corresponding slot. Currently implemented methods are <code>mnn</code>, <code>cca</code>, and <code>combat</code>. MNN is the fastest and will be used here, but it does not generate a corrected gene expression matrix. Therefore, we will use it only to reduce dimentions and perform clustering later on.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="ex">Rscript</span> <span class="va">$script_path</span>/02_integrate.R \</span>
<span id="cb6-2"><a href="#cb6-2"></a>--Seurat_object_path <span class="va">$main</span>/<span class="st">&#39;analysis/1_qc/filt_seurat_object.rds&#39;</span> \</span>
<span id="cb6-3"><a href="#cb6-3"></a>--columns_metadata <span class="va">$var_to_plot</span> \</span>
<span id="cb6-4"><a href="#cb6-4"></a>--regress <span class="va">$var_to_regress</span> \</span>
<span id="cb6-5"><a href="#cb6-5"></a>--var_genes <span class="st">&#39;seurat&#39;</span> \</span>
<span id="cb6-6"><a href="#cb6-6"></a>--integration_method <span class="st">&#39;mnn,dataset&#39;</span> \</span>
<span id="cb6-7"><a href="#cb6-7"></a>--cluster_use <span class="st">&#39;NONE&#39;</span> \</span>
<span id="cb6-8"><a href="#cb6-8"></a>--assay <span class="st">&#39;rna&#39;</span> \</span>
<span id="cb6-9"><a href="#cb6-9"></a>--output_path <span class="va">$main</span>/<span class="st">&#39;analysis/2_clustering&#39;</span> \</span>
<span id="cb6-10"><a href="#cb6-10"></a><span class="op">2&gt;&amp;1</span> <span class="kw">|</span> <span class="fu">tee</span> <span class="va">$main</span>/log/<span class="st">&#39;02_integrate_log.txt&#39;</span></span></code></pre></div>
<p>Now there should be a folder named 2_clustering containig the your seurat object. This object contains everything from the one used as imput plus the integrated slot and variable genes computed.</p>
<p><img src="figures/integration.png" style="width:33.0%" /></p>
<p>This function will then calculate variable genes using both scran and seurat methods for each dataset and for all datasets together. It will output them into the <code>variable_genes</code> folder, which should look like this:</p>
<p><strong>Variable genes folder</strong></p>
<p><img src="figures/variable_gene_folder.png" style="width:50.0%" /></p>
<p><br/></p>
<hr />
</div>
<div id="dim.-reduction-clustering" class="section level1">
<h1>Dim. Reduction &amp; Clustering</h1>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="ex">Rscript</span> <span class="va">$script_path</span>/03_dr_and_cluster.R \</span>
<span id="cb7-2"><a href="#cb7-2"></a>--Seurat_object_path <span class="va">$main</span>/<span class="st">&#39;analysis/2_clustering/seurat_object.rds&#39;</span> \</span>
<span id="cb7-3"><a href="#cb7-3"></a>--columns_metadata <span class="va">$var_to_plot</span> \</span>
<span id="cb7-4"><a href="#cb7-4"></a>--regress <span class="va">$var_to_regress</span> \</span>
<span id="cb7-5"><a href="#cb7-5"></a>--PCs_use <span class="st">&#39;var,1&#39;</span> \</span>
<span id="cb7-6"><a href="#cb7-6"></a>--var_genes <span class="st">&#39;seurat&#39;</span> \</span>
<span id="cb7-7"><a href="#cb7-7"></a>--dim_reduct_use <span class="st">&#39;umap&#39;</span> \</span>
<span id="cb7-8"><a href="#cb7-8"></a>--cluster_use <span class="st">&#39;none&#39;</span> \</span>
<span id="cb7-9"><a href="#cb7-9"></a>--cluster_method <span class="st">&#39;leiden&#39;</span> \</span>
<span id="cb7-10"><a href="#cb7-10"></a>--assay <span class="st">&#39;mnn&#39;</span> \</span>
<span id="cb7-11"><a href="#cb7-11"></a>--output_path <span class="va">$main</span>/<span class="st">&#39;analysis/2_clustering&#39;</span> \</span>
<span id="cb7-12"><a href="#cb7-12"></a><span class="op">2&gt;&amp;1</span> <span class="kw">|</span> <span class="fu">tee</span> <span class="va">$main</span>/log/<span class="st">&#39;03_dr_and_cluster_log.txt&#39;</span></span></code></pre></div>
<p>After running this function, several files and folders should appear in your output folder, which will look something like this:</p>
<p><img src="figures/02_clustering_folder.png" style="width:39.0%" /></p>
<p><strong>PCA folder</strong></p>
<p><img src="figures/pca_folder.png" style="width:45.0%" /></p>
<p><strong>UMAP folder</strong></p>
<p><img src="figures/umap_folder.png" style="width:45.0%" /></p>
<p><strong>Clustering folder</strong></p>
<p><img src="figures/clustering_folder.png" style="width:39.0%" /></p>
</div>
<div id="clustering-analysis" class="section level1">
<h1>Clustering analysis</h1>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">Rscript</span> <span class="va">$script_path</span>/<span class="st">&#39;05_cluster_correlation.R&#39;</span> \</span>
<span id="cb8-2"><a href="#cb8-2"></a>--Seurat_object_path <span class="va">$main</span>/<span class="st">&#39;analysis/2_clustering/seurat_object.rds&#39;</span> \</span>
<span id="cb8-3"><a href="#cb8-3"></a>--clustering_use <span class="st">&#39;HC_12&#39;</span> \</span>
<span id="cb8-4"><a href="#cb8-4"></a>--exclude_cluster <span class="st">&#39;NONE&#39;</span> \</span>
<span id="cb8-5"><a href="#cb8-5"></a>--merge_cluster <span class="st">&#39;0.95,0.9,0.85,0.8,0.75,0.7&#39;</span> \</span>
<span id="cb8-6"><a href="#cb8-6"></a>--output_path <span class="va">$main</span>/<span class="st">&#39;analysis/2_clustering/cluster_correlations&#39;</span> \</span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="op">2&gt;&amp;1</span> <span class="kw">|</span> <span class="fu">tee</span> <span class="va">$main</span>/<span class="st">&#39;log/4_clust_corr.txt&#39;</span></span></code></pre></div>
</div>
<div id="gene-expression-testing" class="section level1">
<h1>Gene expression testing</h1>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="ex">Rscript</span> <span class="va">$script_path</span>/04_diff_gene_expr.R \</span>
<span id="cb9-3"><a href="#cb9-3"></a>--Seurat_object_path <span class="va">$main</span>/<span class="st">&#39;analysis/2_clustering/seurat_object.rds&#39;</span> \</span>
<span id="cb9-4"><a href="#cb9-4"></a>--clustering_use <span class="st">&#39;merged_0.8&#39;</span> \</span>
<span id="cb9-5"><a href="#cb9-5"></a>--metadata_use <span class="st">&#39;dataset&#39;</span> \</span>
<span id="cb9-6"><a href="#cb9-6"></a>--exclude_cluster <span class="st">&#39;NONE&#39;</span> \</span>
<span id="cb9-7"><a href="#cb9-7"></a>--assay <span class="st">&#39;rna&#39;</span> \</span>
<span id="cb9-8"><a href="#cb9-8"></a>--output_path <span class="va">$main</span>/<span class="st">&#39;analysis/2_clustering/diff_expr&#39;</span> \</span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="op">2&gt;&amp;1</span> <span class="kw">|</span> <span class="fu">tee</span> <span class="va">$main</span>/<span class="st">&#39;log/5_diff_expr_log.txt&#39;</span></span></code></pre></div>
</div>
<div id="cell-type-prediction" class="section level1">
<h1>Cell type prediction</h1>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="ex">Rscript</span> <span class="va">$script_path</span>/cell_type_prdiction.R \</span>
<span id="cb10-2"><a href="#cb10-2"></a>--Seurat_object_path <span class="va">$main</span>/<span class="st">&#39;analysis/2_clustering/seurat_object.rds&#39;</span> \</span>
<span id="cb10-3"><a href="#cb10-3"></a>--marker_lists <span class="va">$main</span>/<span class="st">&#39;marer_list.csv&#39;</span> \</span>
<span id="cb10-4"><a href="#cb10-4"></a>--cluster_use <span class="st">&#39;snn_0.4&#39;</span> \</span>
<span id="cb10-5"><a href="#cb10-5"></a>--assay <span class="st">&#39;RNA&#39;</span> \</span>
<span id="cb10-6"><a href="#cb10-6"></a>--output_path <span class="va">$main</span>/<span class="st">&#39;analysis/2_clustering/&#39;</span> \</span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="op">2&gt;&amp;1</span> <span class="kw">|</span> <span class="fu">tee</span> <span class="va">$main</span>/<span class="st">&#39;log/cell_type_prediction_log.txt&#39;</span></span></code></pre></div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
